File: Computer_Science/Algorithm/Learning/Algorithms/二叉树/2. 二叉搜索树/code/98.验证二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=98 lang=cpp
 *
 * [98] 验证二叉搜索树
 */
#include <climits>
#include <utility>
#include <algorithm>
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
using namespace std;

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // 前序遍历
    // bool isValidBST(TreeNode* root, long long left = LLONG_MIN, long long right = LLONG_MAX) {
    //     if (root == nullptr) return true;
    //     long long x = root->val;
    //     return x > left && x < right &&
    //             isValidBST(root->left, left, x) &&
    //             isValidBST(root->right, x, right);
    // }



    // 中序遍历
    // long long pre = LLONG_MIN;
    // bool isValidBST(TreeNode* root) {
    //     if (root == nullptr) return true;
    //     if (!isValidBST(root->left) || root->val <= pre)
    //         return false;
    //     pre = root->val;
    //     return isValidBST(root->right);
    // }

    // 返回值 :
    // 当前这颗树的最小值
    // 当前这颗树的最大值
    pair<long long , long long> dfs(TreeNode* node) {
        if (node == nullptr) return {LLONG_MAX, LLONG_MIN};
        auto[l_min, l_max] = dfs(node->left);
        auto[r_min, r_max] = dfs(node->right);
        long long x = node->val;
        // 不合法的情况
        if (x <= l_max || x >= r_min) return {LLONG_MIN, LLONG_MAX};
        return {min(l_min, x), max(r_max, x)};
    }

    // 后序遍历
    bool isValidBST(TreeNode* root) {
        return dfs(root).second != LLONG_MAX;
    }
};
// @lc code=end

File: Computer_Science/Algorithm/leetcode/hot100/98.验证二叉搜索树.java
/*
 * @lc app=leetcode.cn id=98 lang=java
 *
 * [98] 验证二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public boolean isValidBST(TreeNode root) {
        return isValid(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }

    /**
     * 当前节点的范围是(min, max)
     * @param root
     * @param max 严父对孩子的最大值的限制
     * @param min 对孩子最小值的限制
     * @return
     */
    private boolean isValid(TreeNode root, long min, long max) {
        if (root == null) return true;
        int x = root.val;
        return (x < max && x > min) 
            && isValid(root.left, min, Math.min(max, x)) // (min, min(root, max))
            && isValid(root.right, Math.max(min, x), max); // (min(root, min), max)

    }
}
// @lc code=end

